src/typechecker.rs

```diff
@@ -2156,6 +2189,55 @@ fn typecheck_struct(
     error
 }
 
+fn typecheck_trait(
+    trait_: &ParsedTrait,
+    trait_id: TraitId,
+    parent_scope_id: ScopeId,
+    project: &mut Project,
+) -> Option<JaktError> {
+    let mut error = None;
+
+    let trait_type_id = project.find_or_add_type_id(Type::Trait(trait_id));
+    let trait_scope_id = project.create_scope(parent_scope_id);
+
+    project.traits.push(CheckedTrait {
+        name: trait_.name.clone(),
+        scope_id: trait_scope_id,
+        type_id: trait_type_id,
+    });
+
+    for function in &trait_.methods {
+        error = error.or(typecheck_function_predecl(
+            function,
+            trait_scope_id,
+            Some(trait_type_id),
+            project,
+        ));
+        if error.is_some() {
+            return error;
+        }
+        let function_id = project
+            .find_function_in_scope(trait_scope_id, &function.name)
+            .expect("internal error: function not added to scope");
+        let checked_function = &mut project.functions[function_id];
+        if checked_function.is_static() {
+            // Static functions are not allowed on traits right now, just use namespaces.
```

Comment:
Any reason why? I removed the check and it seemed to be working fine

Reply:
The static function is still on the implementing type, that feels weird to me

Reply:
the trait also expresses requirements on a type, I don't understand why this shouldn't be allowed:
```
trait ArgumentParser {
    function parse(anonymous args: SomeReferenceIterable) throws -> Self
}

struct Foo {
    implements(ArgumentParser) function parse(anonymous args: SomeReferenceIterable) throws -> Foo {
        ...
    }
}
```